server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -413,10 +421,18 @@ private void reroute(RoutingAllocation allocation) {
         assert hasDeadNodes(allocation) == false : "dead nodes should be explicitly cleaned up. See disassociateDeadNodes";
         assert AutoExpandReplicas.getAutoExpandReplicaChanges(allocation.metaData(), allocation).isEmpty() :
             "auto-expand replicas out of sync with number of nodes in the cluster";
+        assert existingShardsAllocators.isEmpty() == false : "must have at least one ExistingShardsAllocator";
 
         removeDelayMarkers(allocation);
+
+        allocation.routingNodes().unassigned().sort(PriorityComparator.getAllocationComparator(allocation)); // sort for priority ordering
+
         // try to allocate existing shard copies first
-        gatewayAllocator.allocateUnassigned(allocation);
+        // TODO TBD the order in which allocators run depends on the order of plugins so is not properly defined. Does this matter?
```

Comment:
for discussion

Reply:
We could leave it to the core framework to decide which allocator to invoke

Reply:
:+1: I've moved to that model now.